packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,377 @@
+  /**
+   * Number of attempts before a peer is considered non-dialable
+   * This is used to not spam a peer with dial attempts when it is not dialable
+   */
+  maxDialAttemptsForPeer?: number;
+  /**
+   * Max number of bootstrap peers allowed to be connected to, initially
+   * This is used to increase intention of dialing non-bootstrap peers, found using other discovery mechanisms (like Peer Exchange)
+   * Is overridden by {@link maxDialAttemptsBeforeBootstrapFallback}
+   */
+  maxBootstrapPeersAllowed?: number;
+  // TODO: add this functionality back
+  // /**
+  //  * Number of attempts before dialling bootstrap nodes over the {@link maxBootstrapPeersAllowed} value.
+  //  * Dialing is the process of opening an outgoing connection to a listening peer with a transport that is supported by both peers.
+  //  * This is only used when other discovery mechanisms haven't yield peers that are dialable
+  //  * This increases relative centralization of the network
+  //  */
+  // maxTimeBeforeBootstrapFallbackMs?: number;
+}
+
+export class ConnectionManager extends KeepAliveManager {
+  private static instances = new Map<string, ConnectionManager>();
+  private options: Options;
+  private libp2pComponents: Libp2pComponents;
+  private dialAttemptsForPeer: Map<string, number> = new Map();
+
+  public static create(
+    peerId: string,
+    libp2p: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ): ConnectionManager {
+    let instance = ConnectionManager.instances.get(peerId);
+    if (!instance) {
+      instance = new ConnectionManager(
+        libp2p,
+        keepAliveOptions,
+        relay,
+        options
+      );
+      ConnectionManager.instances.set(peerId, instance);
+    }
+    return instance;
+  }
+
+  constructor(
+    libp2pComponents: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ) {
+    super();
+    this.libp2pComponents = libp2pComponents;
+    this.options = options ?? {};
+
+    this.runService(keepAliveOptions, relay)
+      .then(() => log(`Connection Manager is now running`))
+      .catch((error) => log(`Unexpected error while running service`, error));
+  }
+
+  private async runService(
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay
+  ): Promise<void> {
+    // log state every N seconds
+    setInterval(async () => {
+      this.fetchUpdatedState();
+    }, DEFAULT_LOGGING_INTERVAL);
+
+    // start event listeners
+    this.startPeerDiscoveryListener();
+    this.startPeerConnectionListener(keepAliveOptions, relay);
+    this.startPeerDisconnectionListener();
+  }
+
+  stopService(): void {
+    this.stopAllKeepAlives();
+    this.libp2pComponents.connectionManager.removeEventListener("peer:connect");
+    this.libp2pComponents.connectionManager.removeEventListener(
+      "peer:disconnect"
+    );
+    this.libp2pComponents.peerStore.removeEventListener("peer");
```

Comment:
If there are guarantees that ConnectionManager will be used this way then your thought is definitely correct, though usually I try to stick to clear-after-yourself approach so that there are no unintended consequences and hard to root cause bugs.